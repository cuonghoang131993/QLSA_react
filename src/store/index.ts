import { useDispatch, useSelector } from "react-redux";
import {
  combineReducers,
  configureStore,
  createDraftSafeSelector,
} from "@reduxjs/toolkit";
import appReducer from "./app/slice";
import userReducer from "./user/slice";
import permissionReducer from "./permission/slice";
import typeReducer from "./type/slice";
import programReducer from "./program/slice";

export const rootReducer = combineReducers({
  app: appReducer,
  user: userReducer,
  permission: permissionReducer,
  type: typeReducer,
  program: programReducer,
});

export function setupStore(preloadedState?: Partial<RootState>): any {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}

// export const store = configureStore({
//   reducer: rootReducer,
// });

export const useAppDispatch = useDispatch.withTypes<AppDispatch>()
export const useAppSelector = useSelector.withTypes<RootState>()

export const createTypedDraftSafeSelector =
  createDraftSafeSelector.withTypes<RootState>();
