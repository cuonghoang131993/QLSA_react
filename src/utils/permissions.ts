export const isRoutePermitted = (
  permissions: string[],
  currentUserPermissions: string[]
) => permissions.every((p) => currentUserPermissions.includes(p));
