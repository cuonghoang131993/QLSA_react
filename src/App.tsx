import React, { Suspense, useEffect } from "react";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import { PATH } from "constants/paths";
import { useAppDispatch } from "store/index";
import { currentUser } from "store/user/action";
import "./styles/styles.css";

const Loading = React.lazy(() => import("components/Loading/Loading"));
const ErrorBoundary = React.lazy(() => import("./ErrorBoundary"));
const MainLayout = React.lazy(() => import("layouts/MainLayout"));
const PrivateRoute = React.lazy(() => import("routes/PrivateRoute"));
const Login = React.lazy(() => import("pages/Login"));

const App = () => {
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(currentUser());
  }, [dispatch]);

  return (
    <ErrorBoundary fallback={<p>Custom fallback for this specific area.</p>}>
      <BrowserRouter>
        <Suspense fallback={<Loading />}>
          <Routes>
            <Route path={PATH.LOGIN} element={<Login />} />
            <Route element={<PrivateRoute />}>
              <Route path="*" element={<MainLayout />} />
            </Route>
          </Routes>
        </Suspense>
      </BrowserRouter>
    </ErrorBoundary>
  );
};

export default App;
