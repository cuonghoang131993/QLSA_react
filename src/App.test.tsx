import { renderWithProviders } from "utils/test-utils";
import { waitFor, screen } from "@testing-library/react";

import App from "./App";
import * as store from "store/index";

const mockNavigate = jest.fn();
// Mock the useRoutes hook
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"), // Import and retain default exports
  useRoutes: jest.fn().mockReturnValue(<div>Mocked Route Content</div>), // Mock the useRoutes function
  useNavigate: () => mockNavigate,
}));
jest.spyOn(store, "useAppDispatch").mockReturnValue(() => {});

describe("App tests", () => {
  it("should contains the heading in login page 1", async () => {
    // Arrange
    renderWithProviders(<App />);

    // Act

    // Assert
    await waitFor(() => {
      const loginText = screen.getAllByText(/Login/i);
      expect(loginText[0]).toHaveClass("text-4xl");
    });
  });
});
