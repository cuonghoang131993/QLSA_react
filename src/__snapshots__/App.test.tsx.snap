// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`App tests should contains the heading 1 1`] = `<div />`;
